### ISSUE TITLE
[FSDP] "reshard_after_forward == False" is equivalent to "ShardedDDP"?

### ISSUE THREAD

**Issue by foreveronehundred** [facebookresearch/fairscale#961]
According to our discussion in the previous issue (https://github.com/facebookresearch/fairscale/issues/943), you mentioned that FSDP with reshard_after_forward == False is equivalent to ShardedDDP. However, I do not think FSDP with this setting is equivalent to ShardedDDP.

In FSDP, we shard module parameters at first and rebuild the parameters during forward. When reshard_after_forward == False, we keep the full parameters after forward, and free them after backward. In other word, we will not always keep the full parameters in FSDP. On the other hand, ShardedDDP does not shard module parameters and always keeps the full parameters.

To sum up, I do not agree with the idea that "FSDP with reshard_after_forward == False" is equivalent to ShardedDDP. Please tell me if my statement is not correct.

**Comment by blefaudeux** [comment]
Hi, thanks for your question, although the wording is a little strange to me. I'll let the FSDP experts chime in, but at the very least you can check out the code base 

> In FSDP, we shard module parameters at first and rebuild the parameters during forward. When reshard_after_forward == False, we keep the full parameters after forward, and free them after backward. 

I think that the code  disagrees with you on that https://github.com/facebookresearch/fairscale/blob/1bc96fa8c69def6d990e42bfbd75f86146ce29bd/fairscale/nn/data_parallel/fully_sharded_data_parallel.py#L1406

https://github.com/facebookresearch/fairscale/blob/1bc96fa8c69def6d990e42bfbd75f86146ce29bd/fairscale/nn/data_parallel/fully_sharded_data_parallel.py#L1633

**Comment by blefaudeux** [comment]
To elaborate a little, the grad sharding is not done in the same way in between FSDP with no_reshard and ShardedDDP, ShardedDDP distributes the parameters around while FSDP is capable of sharding them, and the communication pattern is not the same. It is equivalent in terms of features however (in that the optimizer step is sharded and the gradients are sharded across the fleet), typically called "zero 2", unless a major bug exists (unlikely at this point).

**Comment by foreveronehundred** [comment]
> Hi, thanks for your question, although the wording is a little strange to me. I'll let the FSDP experts chime in, but at the very least you can check out the code base
> 
> > In FSDP, we shard module parameters at first and rebuild the parameters during forward. When reshard_after_forward == False, we keep the full parameters after forward, and free them after backward.
> 
> I think that the code disagrees with you on that
> 
> https://github.com/facebookresearch/fairscale/blob/1bc96fa8c69def6d990e42bfbd75f86146ce29bd/fairscale/nn/data_parallel/fully_sharded_data_parallel.py#L1406
> 
> https://github.com/facebookresearch/fairscale/blob/1bc96fa8c69def6d990e42bfbd75f86146ce29bd/fairscale/nn/data_parallel/fully_sharded_data_parallel.py#L1633

Thanks for the reply. I think the line 1406 and line 1633 **agree** my comment if _require_backward_grad_sync == True.
Let's focus on the case that reshard_after_forward == False. According to line 1405-1408, we will not free full parameters. However, in _post_backward_hook, the full parameters will still be freed if _require_backward_grad_sync == True, and I think _require_backward_grad_sync will be True for every N steps.
To sum up, I think my comment is correct when _require_backward_grad_sync == True in a step

**Comment by blefaudeux** [comment]
cc @tmarkstrum

**Comment by tmarkstrum** [comment]
When we say the reshard_after_forward==False is equivalent to ShardedDDP, we mean the peak memory usage are the same because the GPU does not free full parameters. We did not take _require_backward_grad_sync == True into account when we have the comment.

**Comment by foreveronehundred** [comment]
Thanks for the explanations.

**Comment by blefaudeux** [comment]
> When we say the reshard_after_forward==False is equivalent to ShardedDDP, we mean the peak memory usage are the same because the GPU does not free full parameters. We did not take _require_backward_grad_sync == True into account when we have the comment.

The parameters are re-synced on the next FW then, after the optimizer step I suppose. Then it's not so different from SDDP @foreveronehundred, SharedDDP is not dropping them but it could, since new ones are pushed shard by shard to all the fleet anyway ([see here](https://github.com/facebookresearch/fairscale/blob/main/fairscale/optim/oss.py#L589)). 

I still think that it could be nice to have the option not to drop the parameters (at least if there's some speed attached to that, re-establishing the flat buffers is probably not free), in case one wants to focus on the speed benefits from the sharding, [recent benchmarks from the Pytorch](https://machinelearningnewshubb.com/pytorch-data-parallel-best-practices-on-google-cloud-by-pytorch-mar-2022/) team showed that FSDP could probably be faster (in that it was slower than DDP in a lot of situations, which feels counter intuitive).

In any case this should be better than SDDP @foreveronehundred, because SDDP considers the whole model as a monolith, meaning that this initial shard broadcast part will block the whole forward pass. FSDP breaks the model into submodules and many FWs, which means that this communication part has a higher chance of being shadowed by compute (ie: not wasted time).
